### DISCUSSION TITLE
About Collection<record> in record type.

### DISCUSSION THREAD

**Discussion by PMExtra** [dotnet/runtime#47401] [Q&A]
I read the document: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/records

It said that *The synthesized Equals(R?) returns true if and only if each of the following are true:*
There is a rule that confused me: *the value of System.Collections.Generic.EqualityComparer<TN>.Default.Equals(fieldN, other.fieldN) where TN is the field type*

In my case, I have a record type like:
```cs
public record Item { ... }
public record Order {
    public List<Item> Items { get; init; }
}
```
Then I want to compare between two Order instances, that fails. I think it because record type is not field.

So why don't the records make sequence equal compare in Collection?

**Comment by Clockwork-Muse** [comment]
Probably because you can't do sequence-based equal for everything that implements `IEnumerable<whatever>` - Consider the case of `Dictionary` or `Set`, where the contents may be equal (keys and their values), but the order returned by enumerating them would be different.  This leaves calling `Equals`, which I believe in most cases is only the default one (ie, "compare object identity"), since a full compare can be expensive.

**Reply by PMExtra** [reply]
But it's record type.
A reference basis is that, the `@Data` class in Java (lombok) will regard the objects equal if the collection properties are sequence equal.

**Reply by PathogenDavid** [reply]
This is really a discussion for [csharplang](https://github.com/dotnet/csharplang).

You can find the language design meeting notes from when they decided records would only use shallow quality in the [2020/03/30 meeting notes](https://github.com/dotnet/csharplang/blob/6ab8409a32f1249f9aef618054426acb7bc7b4c1/meetings/2020/LDM-2020-03-30.md#value-equality). (The [2020/02/26 meeting notes](https://github.com/dotnet/csharplang/blob/6ab8409a32f1249f9aef618054426acb7bc7b4c1/meetings/2020/LDM-2020-02-26.md#value-equality) also mention it.)

The short version is that they were concerned that adding anything more complex than basic shallow value equality would add too many extension points and thought source generators would be a better option for most customizable equality implementations.

> A reference basis is that, the `@Data` class in Java (lombok) will regard the objects equal if the collection properties are sequence equal.

Java has set a different precedent for collection equality in general. For instance, Java's [`AbstractList.equals`](https://docs.oracle.com/javase/7/docs/api/java/util/AbstractList.html#equals(java.lang.Object)) method provides sequence equality semantics. .NET's `List<T>` does not.

**Reply by PMExtra** [reply]
@PathogenDavid Thank you very much. You solved my confusion.

**Reply by PathogenDavid** [reply]
No problem, glad I could help!
